patches/server/0977-Add-event-for-player-editing-sign.patch

```diff
@@ -0,0 +1,53 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: by77er <by77er@gmail.com>
+Date: Sat, 10 Jun 2023 19:09:03 -0400
+Subject: [PATCH] Add event for player editing sign
+
+
+diff --git a/src/main/java/net/minecraft/world/level/block/SignBlock.java b/src/main/java/net/minecraft/world/level/block/SignBlock.java
+index 91bb294be2cd8ab3467a62006a5a2751e0bec4ba..0f02bd1acaf12be255958ddf27d8c7993e15e873 100644
+--- a/src/main/java/net/minecraft/world/level/block/SignBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/SignBlock.java
+@@ -76,9 +76,9 @@ public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterlo
+     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
+         ItemStack itemStack = player.getItemInHand(hand);
+         Item item = itemStack.getItem();
+-        Item signBlockEntity = itemStack.getItem();
++        Item potentialApplicator = itemStack.getItem(); // Paper: renamed signBlockEntity -> potentialApplicator due to decompile error
+         SignApplicator var10000;
+-        if (signBlockEntity instanceof SignApplicator signApplicator) {
++        if (potentialApplicator instanceof SignApplicator signApplicator) { // Paper: renamed signBlockEntity -> potentialApplicator due to decompile error
+             var10000 = signApplicator;
+         } else {
+             var10000 = null;
+@@ -86,8 +86,8 @@ public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterlo
+ 
+         SignApplicator signApplicator2 = var10000;
+         boolean bl = signApplicator2 != null && player.mayBuild();
+-        BlockEntity bl2 = world.getBlockEntity(pos);
+-        if (bl2 instanceof SignBlockEntity signBlockEntity) {
++        BlockEntity blockEntity = world.getBlockEntity(pos); // Paper: renamed bl2 -> blockEntity due to decompile error
++        if (blockEntity instanceof SignBlockEntity signBlockEntity) { // Paper: renamed bl2 -> blockEntity due to decompile error
+             if (!world.isClientSide) {
+                 boolean bl2 = signBlockEntity.isFacingFrontText(player);
+                 SignText signText = signBlockEntity.getText(bl2);
+@@ -153,8 +153,17 @@ public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterlo
+     }
+ 
+     public void openTextEdit(Player player, SignBlockEntity blockEntity, boolean front) {
+-        blockEntity.setAllowedPlayerEditor(player.getUUID());
+-        player.openTextEdit(blockEntity, front);
++        // Paper start
++        var bukkitPlayer = (org.bukkit.entity.Player) player.getBukkitEntity();
++        var event = new com.destroystokyo.paper.event.player.PlayerEditSignEvent(
++            bukkitPlayer,
++            new org.bukkit.craftbukkit.block.CraftSign<>(bukkitPlayer.getWorld(), blockEntity),
```

Comment:
This isn't correct, it could be a hanging sign. CraftSign doesn't cover that. You should instead use one of the methods on `CraftBlockStates` to get the blockstate instance.

Reply:
In working on this I realized it might be better to return a `Block` in place of the `Sign`, since the event handler probably needs more context anyway. Thoughts?

Reply:
Sign is fine as HangingSign extends it, but you need to create a CraftHangingSign if its a hanging sign instead of CraftSign for a regular sign.

Reply:
Alright, let me know if it looks good now.